File: samples/CameraAccess/CameraAccess/WebRTC/WebRTCSessionViewModel.swift
import Foundation
import SwiftUI
import WebRTC

enum WebRTCConnectionState: Equatable {
  case disconnected
  case connecting
  case waitingForPeer
  case connected
  case backgrounded
  case error(String)
}

/// Orchestrates the WebRTC live streaming session: signaling, peer connection, and frame forwarding.
/// Follows the same @MainActor ObservableObject pattern as GeminiSessionViewModel.
@MainActor
class WebRTCSessionViewModel: ObservableObject {
  @Published var isActive: Bool = false
  @Published var connectionState: WebRTCConnectionState = .disconnected
  @Published var roomCode: String = ""
  @Published var isMuted: Bool = false
  @Published var errorMessage: String?
  @Published var remoteVideoTrack: RTCVideoTrack?
  @Published var hasRemoteVideo: Bool = false

  private var webRTCClient: WebRTCClient?
  private var signalingClient: SignalingClient?
  private var delegateAdapter: WebRTCDelegateAdapter?

  /// Saved room code for reconnecting after app backgrounding.
  private var savedRoomCode: String?
  private var foregroundObserver: Any?

  func startSession() async {
    guard !isActive else { return }
    guard WebRTCConfig.isConfigured else {
      errorMessage = "WebRTC signaling URL not configured."
      return
    }

    isActive = true
    connectionState = .connecting
    savedRoomCode = nil

    // Fetch TURN credentials for NAT traversal across networks
    let iceServers = await WebRTCConfig.fetchIceServers()

    setupWebRTCClient(iceServers: iceServers)
    connectSignaling(rejoinCode: nil)
    observeForeground()
  }

  func stopSession() {
    removeForegroundObserver()
    webRTCClient?.close()
    webRTCClient = nil
    delegateAdapter = nil
    signalingClient?.disconnect()
    signalingClient = nil
    isActive = false
    connectionState = .disconnected
    roomCode = ""
    savedRoomCode = nil
    isMuted = false
    remoteVideoTrack = nil
    hasRemoteVideo = false
  }

  func toggleMute() {
    isMuted.toggle()
    webRTCClient?.muteAudio(isMuted)
  }

  /// Called by StreamSessionViewModel on each video frame.
  func pushVideoFrame(_ image: UIImage) {
    guard isActive, connectionState == .connected else { return }
    webRTCClient?.pushVideoFrame(image)
  }

  // MARK: - WebRTC + Signaling Setup

  private func setupWebRTCClient(iceServers: [RTCIceServer]?) {
    let client = WebRTCClient()
    let adapter = WebRTCDelegateAdapter(viewModel: self)
    delegateAdapter = adapter
    client.delegate = adapter
    client.setup(iceServers: iceServers)
    webRTCClient = client
  }

  private func connectSignaling(rejoinCode: String?) {
    signalingClient?.disconnect()

    let signaling = SignalingClient()
    signalingClient = signaling

    signaling.onConnected = { [weak self] in
      Task { @MainActor in
        if let code = rejoinCode {
          NSLog("[WebRTC] Reconnected, rejoining room: %@", code)
          self?.signalingClient?.rejoinRoom(code: code)
        } else {
          self?.signalingClient?.createRoom()
        }
      }
    }

    signaling.onMessageReceived = { [weak self] message in
      Task { @MainActor in
        self?.handleSignalingMessage(message)
      }
    }

    signaling.onDisconnected = { [weak self] reason in
      Task { @MainActor in
        guard let self, self.isActive else { return }
        // Don't fully stop -- mark as backgrounded so we can reconnect
        if self.savedRoomCode != nil {
          self.connectionState = .backgrounded
          NSLog("[WebRTC] Signaling disconnected (backgrounded), will rejoin: %@", reason ?? "unknown")
        } else {
          self.stopSession()
          self.errorMessage = "Signaling disconnected: \(reason ?? "Unknown")"
        }
      }
    }

    guard let url = URL(string: WebRTCConfig.signalingServerURL) else {
      errorMessage = "Invalid signaling URL"
      isActive = false
      connectionState = .disconnected
      return
    }
    signaling.connect(url: url)
  }

  // MARK: - Foreground Reconnect

  private func observeForeground() {
    removeForegroundObserver()
    foregroundObserver = NotificationCenter.default.addObserver(
      forName: UIApplication.willEnterForegroundNotification,
      object: nil,
      queue: .main
    ) { [weak self] _ in
      Task { @MainActor in
        self?.handleReturnToForeground()
      }
    }
  }

  private func removeForegroundObserver() {
    if let observer = foregroundObserver {
      NotificationCenter.default.removeObserver(observer)
      foregroundObserver = nil
    }
  }

  private func handleReturnToForeground() {
    guard isActive, let code = savedRoomCode else { return }
    NSLog("[WebRTC] App returned to foreground, reconnecting to room: %@", code)
    connectionState = .connecting

    // Tear down old peer connection, set up fresh one
    webRTCClient?.close()
    remoteVideoTrack = nil
    hasRemoteVideo = false

    Task {
      let iceServers = await WebRTCConfig.fetchIceServers()
      setupWebRTCClient(iceServers: iceServers)
      connectSignaling(rejoinCode: code)
    }
  }

  // MARK: - Signaling Message Handling

  private func handleSignalingMessage(_ message: SignalingMessage) {
    switch message {
    case .roomCreated(let code):
      roomCode = code
      savedRoomCode = code
      connectionState = .waitingForPeer
      NSLog("[WebRTC] Room created: %@", code)

    case .roomRejoined(let code):
      roomCode = code
      savedRoomCode = code
      connectionState = .waitingForPeer
      NSLog("[WebRTC] Room rejoined: %@", code)

    case .peerJoined:
      NSLog("[WebRTC] Peer joined, creating offer")
      webRTCClient?.createOffer { [weak self] sdp in
        self?.signalingClient?.send(sdp: sdp)
      }

    case .answer(let sdp):
      webRTCClient?.set(remoteSdp: sdp) { error in
        if let error {
          NSLog("[WebRTC] Error setting remote SDP: %@", error.localizedDescription)
        }
      }

    case .candidate(let candidate):
      webRTCClient?.set(remoteCandidate: candidate) { error in
        if let error {
          NSLog("[WebRTC] Error adding ICE candidate: %@", error.localizedDescription)
        }
      }

    case .peerLeft:
      NSLog("[WebRTC] Peer left")
      connectionState = .waitingForPeer

    case .error(let msg):
      // If rejoin fails (room expired), fall back to creating a new room
      if savedRoomCode != nil && msg == "Room not found" {
        NSLog("[WebRTC] Rejoin failed (room expired), creating new room")
        savedRoomCode = nil
        signalingClient?.createRoom()
      } else {
        errorMessage = msg
      }

    case .roomJoined, .offer:
      break
    }
  }

  // MARK: - Connection State Updates (from WebRTCClient delegate)

  fileprivate func handleConnectionStateChange(_ state: RTCIceConnectionState) {
    switch state {
    case .connected, .completed:
      connectionState = .connected
      NSLog("[WebRTC] Peer connected")
    case .disconnected:
      connectionState = .waitingForPeer
    case .failed:
      connectionState = .error("Connection failed")
    case .closed:
      connectionState = .disconnected
    default:
      break
    }
  }

  fileprivate func handleGeneratedCandidate(_ candidate: RTCIceCandidate) {
    signalingClient?.send(candidate: candidate)
  }

  fileprivate func handleRemoteVideoTrackReceived(_ track: RTCVideoTrack) {
    remoteVideoTrack = track
    hasRemoteVideo = true
    NSLog("[WebRTC] Remote video track received")
  }

  fileprivate func handleRemoteVideoTrackRemoved(_ track: RTCVideoTrack) {
    remoteVideoTrack = nil
    hasRemoteVideo = false
    NSLog("[WebRTC] Remote video track removed")
  }
}

// MARK: - Delegate Adapter (bridges nonisolated delegate to @MainActor ViewModel)

private class WebRTCDelegateAdapter: WebRTCClientDelegate {
  private weak var viewModel: WebRTCSessionViewModel?

  init(viewModel: WebRTCSessionViewModel) {
    self.viewModel = viewModel
  }

  func webRTCClient(_ client: WebRTCClient, didChangeConnectionState state: RTCIceConnectionState) {
    Task { @MainActor [weak self] in
      self?.viewModel?.handleConnectionStateChange(state)
    }
  }

  func webRTCClient(_ client: WebRTCClient, didGenerateCandidate candidate: RTCIceCandidate) {
    Task { @MainActor [weak self] in
      self?.viewModel?.handleGeneratedCandidate(candidate)
    }
  }

  func webRTCClient(_ client: WebRTCClient, didReceiveRemoteVideoTrack track: RTCVideoTrack) {
    Task { @MainActor [weak self] in
      self?.viewModel?.handleRemoteVideoTrackReceived(track)
    }
  }

  func webRTCClient(_ client: WebRTCClient, didRemoveRemoteVideoTrack track: RTCVideoTrack) {
    Task { @MainActor [weak self] in
      self?.viewModel?.handleRemoteVideoTrackRemoved(track)
    }
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/webrtc/WebRTCSessionViewModel.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.webrtc

import android.app.Application
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.VideoTrack

sealed class WebRTCConnectionState {
    object Disconnected : WebRTCConnectionState()
    object Connecting : WebRTCConnectionState()
    object WaitingForPeer : WebRTCConnectionState()
    object Connected : WebRTCConnectionState()
    object Backgrounded : WebRTCConnectionState()
    data class Error(val message: String) : WebRTCConnectionState()
}

data class WebRTCUiState(
    val isActive: Boolean = false,
    val connectionState: WebRTCConnectionState = WebRTCConnectionState.Disconnected,
    val roomCode: String = "",
    val isMuted: Boolean = false,
    val errorMessage: String? = null,
    val remoteVideoTrack: VideoTrack? = null,
    val hasRemoteVideo: Boolean = false,
)

class WebRTCSessionViewModel(application: Application) : AndroidViewModel(application) {
    companion object {
        private const val TAG = "WebRTCSessionVM"
    }

    private val _uiState = MutableStateFlow(WebRTCUiState())
    val uiState: StateFlow<WebRTCUiState> = _uiState.asStateFlow()

    private var webRTCClient: WebRTCClient? = null
    private var signalingClient: SignalingClient? = null
    private var savedRoomCode: String? = null

    private val lifecycleObserver = object : DefaultLifecycleObserver {
        override fun onStart(owner: LifecycleOwner) {
            // App came to foreground
            handleReturnToForeground()
        }
    }

    fun startSession() {
        if (_uiState.value.isActive) return

        if (!WebRTCConfig.isConfigured) {
            _uiState.value = _uiState.value.copy(
                errorMessage = "WebRTC signaling URL not configured."
            )
            return
        }

        _uiState.value = _uiState.value.copy(
            isActive = true,
            connectionState = WebRTCConnectionState.Connecting,
        )
        savedRoomCode = null

        viewModelScope.launch {
            val iceServers = WebRTCConfig.fetchIceServers()
            setupWebRTCClient(iceServers)
            connectSignaling(rejoinCode = null)
            observeForeground()
        }
    }

    fun stopSession() {
        removeForegroundObserver()
        webRTCClient?.close()
        webRTCClient = null
        signalingClient?.disconnect()
        signalingClient = null
        savedRoomCode = null
        _uiState.value = WebRTCUiState()
    }

    fun toggleMute() {
        val newMuted = !_uiState.value.isMuted
        _uiState.value = _uiState.value.copy(isMuted = newMuted)
        webRTCClient?.muteAudio(newMuted)
    }

    fun copyRoomCode() {
        val code = _uiState.value.roomCode
        if (code.isNotEmpty()) {
            val clipboard = getApplication<Application>()
                .getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            clipboard.setPrimaryClip(ClipData.newPlainText("Room Code", code))
        }
    }

    fun pushVideoFrame(bitmap: Bitmap) {
        if (!_uiState.value.isActive) return
        if (_uiState.value.connectionState != WebRTCConnectionState.Connected) return
        webRTCClient?.pushVideoFrame(bitmap)
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    // -- Private --

    private fun setupWebRTCClient(iceServers: List<PeerConnection.IceServer>) {
        webRTCClient?.close()
        val client = WebRTCClient(getApplication())
        client.delegate = object : WebRTCClientDelegate {
            override fun onConnectionStateChanged(state: PeerConnection.IceConnectionState) {
                viewModelScope.launch { handleConnectionStateChange(state) }
            }
            override fun onIceCandidateGenerated(candidate: IceCandidate) {
                signalingClient?.sendCandidate(candidate)
            }
            override fun onRemoteVideoTrackReceived(track: VideoTrack) {
                viewModelScope.launch {
                    _uiState.value = _uiState.value.copy(
                        remoteVideoTrack = track,
                        hasRemoteVideo = true,
                    )
                    Log.d(TAG, "Remote video track received")
                }
            }
            override fun onRemoteVideoTrackRemoved(track: VideoTrack) {
                viewModelScope.launch {
                    _uiState.value = _uiState.value.copy(
                        remoteVideoTrack = null,
                        hasRemoteVideo = false,
                    )
                    Log.d(TAG, "Remote video track removed")
                }
            }
        }
        client.setup(iceServers)
        webRTCClient = client
    }

    private fun connectSignaling(rejoinCode: String?) {
        signalingClient?.disconnect()

        val signaling = SignalingClient()
        signalingClient = signaling

        signaling.onConnected = {
            viewModelScope.launch {
                if (rejoinCode != null) {
                    Log.d(TAG, "Reconnected, rejoining room: $rejoinCode")
                    signalingClient?.rejoinRoom(rejoinCode)
                } else {
                    signalingClient?.createRoom()
                }
            }
        }

        signaling.onMessageReceived = { message ->
            viewModelScope.launch { handleSignalingMessage(message) }
        }

        signaling.onDisconnected = { reason ->
            viewModelScope.launch {
                if (!_uiState.value.isActive) return@launch
                if (savedRoomCode != null) {
                    _uiState.value = _uiState.value.copy(
                        connectionState = WebRTCConnectionState.Backgrounded
                    )
                    Log.d(TAG, "Signaling disconnected (backgrounded), will rejoin: $reason")
                } else {
                    stopSession()
                    _uiState.value = _uiState.value.copy(
                        errorMessage = "Signaling disconnected: ${reason ?: "Unknown"}"
                    )
                }
            }
        }

        signaling.connect(WebRTCConfig.signalingServerURL)
    }

    private fun handleSignalingMessage(message: SignalingMessage) {
        when (message) {
            is SignalingMessage.RoomCreated -> {
                _uiState.value = _uiState.value.copy(
                    roomCode = message.room,
                    connectionState = WebRTCConnectionState.WaitingForPeer,
                )
                savedRoomCode = message.room
                Log.d(TAG, "Room created: ${message.room}")
            }
            is SignalingMessage.RoomRejoined -> {
                _uiState.value = _uiState.value.copy(
                    roomCode = message.room,
                    connectionState = WebRTCConnectionState.WaitingForPeer,
                )
                savedRoomCode = message.room
                Log.d(TAG, "Room rejoined: ${message.room}")
            }
            is SignalingMessage.PeerJoined -> {
                Log.d(TAG, "Peer joined, creating offer")
                webRTCClient?.createOffer { sdp ->
                    signalingClient?.sendSdp(sdp)
                }
            }
            is SignalingMessage.Answer -> {
                webRTCClient?.setRemoteSdp(message.sdp) { error ->
                    error?.let { Log.e(TAG, "Error setting remote SDP: $it") }
                }
            }
            is SignalingMessage.Candidate -> {
                webRTCClient?.addRemoteCandidate(message.candidate) { error ->
                    error?.let { Log.e(TAG, "Error adding ICE candidate: $it") }
                }
            }
            is SignalingMessage.PeerLeft -> {
                Log.d(TAG, "Peer left")
                _uiState.value = _uiState.value.copy(
                    connectionState = WebRTCConnectionState.WaitingForPeer,
                )
            }
            is SignalingMessage.Error -> {
                // If rejoin fails (room expired), create a new room
                if (savedRoomCode != null && message.message == "Room not found") {
                    Log.d(TAG, "Rejoin failed (room expired), creating new room")
                    savedRoomCode = null
                    signalingClient?.createRoom()
                } else {
                    _uiState.value = _uiState.value.copy(errorMessage = message.message)
                }
            }
            is SignalingMessage.RoomJoined, is SignalingMessage.Offer -> {
                // Not handled by streamer side
            }
        }
    }

    private fun handleConnectionStateChange(state: PeerConnection.IceConnectionState) {
        when (state) {
            PeerConnection.IceConnectionState.CONNECTED,
            PeerConnection.IceConnectionState.COMPLETED -> {
                _uiState.value = _uiState.value.copy(
                    connectionState = WebRTCConnectionState.Connected
                )
                Log.d(TAG, "Peer connected")
            }
            PeerConnection.IceConnectionState.DISCONNECTED -> {
                _uiState.value = _uiState.value.copy(
                    connectionState = WebRTCConnectionState.WaitingForPeer
                )
            }
            PeerConnection.IceConnectionState.FAILED -> {
                _uiState.value = _uiState.value.copy(
                    connectionState = WebRTCConnectionState.Error("Connection failed")
                )
            }
            PeerConnection.IceConnectionState.CLOSED -> {
                _uiState.value = _uiState.value.copy(
                    connectionState = WebRTCConnectionState.Disconnected
                )
            }
            else -> {}
        }
    }

    private fun handleReturnToForeground() {
        val code = savedRoomCode ?: return
        if (!_uiState.value.isActive) return

        Log.d(TAG, "App returned to foreground, reconnecting to room: $code")
        _uiState.value = _uiState.value.copy(
            connectionState = WebRTCConnectionState.Connecting,
            remoteVideoTrack = null,
            hasRemoteVideo = false,
        )

        webRTCClient?.close()

        viewModelScope.launch {
            val iceServers = WebRTCConfig.fetchIceServers()
            setupWebRTCClient(iceServers)
            connectSignaling(rejoinCode = code)
        }
    }

    private fun observeForeground() {
        ProcessLifecycleOwner.get().lifecycle.addObserver(lifecycleObserver)
    }

    private fun removeForegroundObserver() {
        ProcessLifecycleOwner.get().lifecycle.removeObserver(lifecycleObserver)
    }

    override fun onCleared() {
        super.onCleared()
        stopSession()
    }
}
